aptos-move/dynamic-transaction-composer/src/decompiler.rs

```diff
@@ -0,0 +1,279 @@
+}
+
+struct CallState {
+    args: Vec<BatchArgument>,
+    call_info: Option<(ModuleId, Identifier, Vec<TypeTag>)>,
+    num_returns: u16,
+}
+
+impl LocalState {
+    fn new() -> Self {
+        Self {
+            local_idx_to_return_idx: BTreeMap::new(),
+            payload: vec![],
+            current_call: CallState {
+                args: vec![],
+                call_info: None,
+                num_returns: 0,
+            },
+        }
+    }
+
+    fn process_script(&mut self, script: &CompiledScript, args: &[Vec<u8>]) -> anyhow::Result<()> {
+        let mut op_codes = script.code.code.iter().peekable();
+        while !(matches!(op_codes.peek(), Some(Bytecode::Ret))) {
+            while !matches!(
+                op_codes.peek(),
+                Some(Bytecode::Call(_)) | Some(Bytecode::CallGeneric(_))
+            ) {
+                let op_code = op_codes
+                    .next()
+                    .ok_or_else(|| anyhow!("Unexpected end of opcode sequence"))?
+                    .clone();
+                self.process_input_argument(op_code, script, args)?;
+            }
+            let op_code = op_codes
+                .next()
+                .ok_or_else(|| anyhow!("Unexpected end of opcode sequence"))?
+                .clone();
+            self.process_call(op_code, script, args)?;
+            while matches!(op_codes.peek(), Some(Bytecode::StLoc(_))) {
+                let op_code = op_codes
+                    .next()
+                    .ok_or_else(|| anyhow!("Unexpected end of opcode sequence"))?
+                    .clone();
+                self.process_return(op_code, script, args)?;
+            }
+            self.add_new_call()?;
+        }
+        Ok(())
+    }
+
+    fn add_new_call(&mut self) -> anyhow::Result<()> {
+        let mut new_state = CallState {
+            args: vec![],
+            call_info: None,
+            num_returns: 0,
+        };
+        std::mem::swap(&mut new_state, &mut self.current_call);
+        let (module, function, ty_args) = new_state
+            .call_info
+            .ok_or_else(|| anyhow!("Call information not available"))?;
+        self.payload.push(BatchedFunctionCall {
+            module,
+            function,
+            ty_args,
+            args: new_state.args,
+        });
+        Ok(())
+    }
+
+    fn process_input_argument(
+        &mut self,
+        op: Bytecode,
+        script: &CompiledScript,
+        args: &[Vec<u8>],
+    ) -> anyhow::Result<()> {
+        let (operation_type, local_idx) = match op {
+            Bytecode::MoveLoc(i) => (ArgumentOperation::Move, i),
+            Bytecode::CopyLoc(i) => (ArgumentOperation::Copy, i),
+            Bytecode::ImmBorrowLoc(i) => (ArgumentOperation::Borrow, i),
+            Bytecode::MutBorrowLoc(i) => (ArgumentOperation::BorrowMut, i),
+            _ => bail!("Unexpected opcode: {:?}", op),
+        };
+        if local_idx as usize >= script.signature_at(script.parameters).0.len() {
```

Comment:
Hmm... I have a crazy idea: given the power of the script builder you are exposing already, would it make sense for us to just ban script arguments? I mean if they want to customize them they can do it with the script builder directly...

Signers are tricky though...

Anyway, just something for us to think about.

Reply:
See my previous comment about script caching. One nice thing about having script arguments is you would be able to cache script that only have different paramter types, which could make VM life a lot easier.

Reply:
> you would be able to cache script that only have different paramter types

Very good point.